feat: introduce pInfo.UnschedulableCount to make the backoff calculation more appropriate

sanposhiho opened the pull request:
<!--  Thanks for sending a pull request!  Here are some tips for you:

1. If this is your first time, please read our contributor guidelines: https://git.k8s.io/community/contributors/guide/first-contribution.md#your-first-contribution and developer guide https://git.k8s.io/community/contributors/devel/development.md#development-guide
2. Please label this pull request according to what type of issue you are addressing, especially if this is a release targeted pull request. For reference on required PR/issue labels, read here:
https://git.k8s.io/community/contributors/devel/sig-release/release.md#issuepr-kind-label
3. Ensure you have added or ran the appropriate tests for your PR: https://git.k8s.io/community/contributors/devel/sig-testing/testing.md
4. If you want *faster* PR reviews, read how: https://git.k8s.io/community/contributors/guide/pull-requests.md#best-practices-for-faster-reviews
5. If the PR is unfinished, see how to mark it: https://git.k8s.io/community/contributors/guide/pull-requests.md#marking-unfinished-pull-requests
-->

#### What type of PR is this?


/kind feature

<!--
Add one of the following kinds:
/kind bug
/kind cleanup
/kind documentation
/kind feature

Optionally add one or more of the following kinds if applicable:
/kind api-change
/kind deprecation
/kind failing-test
/kind flake
/kind regression
-->

#### What this PR does / why we need it:

introduce pInfo.UnschedulableCount to make the backoff calculation more appropriate, see #128744 for the argument.

#### Which issue(s) this PR fixes:
<!--
*Automatically closes linked issue when PR is merged.
Usage: `Fixes #<issue number>`, or `Fixes (paste link of issue)`.
_If PR is about `failing-tests or flakes`, please post the related issues/tests in a comment and do not use `Fixes`_*
-->
Fixes #128744

#### Special notes for your reviewer:

#### Does this PR introduce a user-facing change?
<!--
If no, just write "NONE" in the release-note block below.
If yes, a release note is required:
Enter your extended release note in the block below. If the PR requires additional action from users switching to the new release, include the string "action required".

For more information on release notes see: https://git.k8s.io/community/contributors/guide/release-notes.md
-->
```release-note
Non-scheduling related errors (e.g., network errors) don't lengthen the Pod scheduling backoff time.
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

<!--
This section can be blank if this pull request does not require a release note.

When adding links which point to resources within git repositories, like
KEPs or supporting documentation, please reference a specific commit and avoid
linking directly to the master branch. This ensures that links reference a
specific point in time, rather than a document that may change over time.

See here for guidance on getting permanent links to files: https://help.github.com/en/articles/getting-permanent-links-to-files

Please use the following format for linking documentation:
- [KEP]: <link>
- [Usage]: <link>
- [Other doc]: <link>
-->
```docs

```

sanposhiho committed:
feat: introduce pInfo.UnschedulableCount to make the backoff calculation more appropriate

sanposhiho commented:
/cc @macsko @dom4ha

sanposhiho commented:
/hold

To go thru the approver's review as well

haosdent commented on pkg/scheduler/backend/queue/scheduling_queue.go:
```diff
@@ -771,6 +774,11 @@ func (p *PriorityQueue) AddUnschedulableIfNotPresent(logger klog.Logger, pInfo *
 	}
 
+	if pInfo.UnschedulablePlugins.Union(pInfo.PendingPlugins).Len() != 0 {
+		// This Pod is rejected by some plugins, not coming back due to unexpected errors (e.g., a network issue)
```
From the description in https://github.com/kubernetes/kubernetes/issues/128744

```
* Unschedulable: the Pod is unschedulable; e.g., no node has enough resource, nodeaffinity, etc.
* Error: the Pod cannot be scheduled because of unexpected errors; e.g., kube-apiserver is unstable, network issue, etc.
```

UnschedulableCount should not increase when `a network issue`?

sanposhiho replied:
Should not, and here describes/implements that, right?

haosdent replied:
OK, the comment "rejected by some plugins" here makes me think some plugins refer to  UnschedulablePlugins/PendingPlugins.

haosdent commented on pkg/scheduler/backend/queue/scheduling_queue_test.go:
```diff
@@ -1651,14 +1668,16 @@ func TestPriorityQueue_MoveAllToActiveOrBackoffQueueWithQueueingHint(t *testing.
 	}{
 		{
-			name:      "Queue queues pod to activeQ",
+			name: "Queue queues pod to activeQ",
```
Should we fix indent at here?

sanposhiho replied:
Yes. gofmt does this.

haosdent replied:
Got it

haosdent commented on pkg/scheduler/framework/types.go:
```diff
@@ -363,9 +363,14 @@ type QueuedPodInfo struct {
+	// Number of all schedule attempts before successfully scheduled.
+	// It's used to record the # attempts metric.
 	Attempts int
+	// Number of the scheduling attemts that this Pod gets unschedulable.
```
Seems conflict with the comment above

```
		// This Pod is rejected by some plugins, not coming back due to unexpected errors (e.g., a network issue)
```

sanposhiho replied:
How does it conflict?
We increment `UnschedulableCount` when `pInfo.UnschedulablePlugins.Union(pInfo.PendingPlugins).Len() != 0`, which is when the pod is coming back because it gets unschedulable, not coming back due to unexpected errors such as networking issues.

haosdent replied:
OK. I was confused by "by some plugins" before, but after rechecking the comments, it is clear to me now.

macsko commented on pkg/scheduler/framework/types.go:
```diff
@@ -363,9 +363,14 @@ type QueuedPodInfo struct {
+	// Number of all schedule attempts before successfully scheduled.
+	// It's used to record the # attempts metric.
 	Attempts int
+	// Number of the scheduling attemts that this Pod gets unschedulable.
```
```suggestion
	// Number of the scheduling attempts that this Pod gets unschedulable.
```

macsko reviewed (commented):
One nit

/lgtm

sanposhiho commented:
will check the integration test failure..

sanposhiho commented:
/retest

sanposhiho commented on test/integration/scheduler/plugins/plugins_test.go:
```diff
@@ -475,6 +482,12 @@ func (bp *BindPlugin) Bind(ctx context.Context, state *framework.CycleState, p *
 
 	bp.numBindCalled++
 	if bp.pluginInvokeEventChan != nil {
+		defer func() {
```
These recover is necessary because we close the channel after all the test is done and we could hit `send on closed channel` panic.

sanposhiho commented on test/integration/scheduler/plugins/plugins_test.go:
```diff
@@ -1670,6 +1689,7 @@ func TestBindPlugin(t *testing.T) {
 			defer teardown()
 
 			pluginInvokeEventChan = make(chan pluginInvokeEvent, 10)
+			defer close(pluginInvokeEventChan)
```
need to close the channel to let all the plugins know the test is complete.
otherwise we'd hit `found unexpected goroutines` error, which tell us there's some in-progress goroutine.

sanposhiho commented:
/retest

macsko commented on test/integration/scheduler/plugins/plugins_test.go:
```diff
@@ -475,6 +482,12 @@ func (bp *BindPlugin) Bind(ctx context.Context, state *framework.CycleState, p *
 
 	bp.numBindCalled++
 	if bp.pluginInvokeEventChan != nil {
+		defer func() {
+			if r := recover(); r != nil {
+				klog.FromContext(ctx).Error(nil, "Recovered from panic in Bind", "error", r)
+				return
+			}
+		}()
 		bp.pluginInvokeEventChan <- pluginInvokeEvent{pluginName: bp.Name(), val: bp.numBindCalled}
```
Isn't it sufficient to use `select` with `ctx.Done()` instead?
```suggestion
		select {
		case <-ctx.Done():
		case bp.pluginInvokeEventChan <- pluginInvokeEvent{pluginName: bp.Name(), val: bp.numBindCalled}:
		}
```

macsko commented:
/lgtm

sanposhiho commented:
@dom4ha Can you review this one as well, as the second reviewing path.

dom4ha commented on pkg/scheduler/backend/queue/scheduling_queue.go:
```diff
@@ -736,9 +736,12 @@ func (p *PriorityQueue) addUnschedulableWithoutQueueingHint(logger klog.Logger,
-		p.podBackoffQ.AddOrUpdate(pInfo)
-		logger.V(5).Info("Pod moved to an internal scheduling queue", "pod", klog.KObj(pod), "event", framework.ScheduleAttemptFailure, "queue", backoffQ)
-		metrics.SchedulerQueueIncomingPods.WithLabelValues("backoff", framework.ScheduleAttemptFailure).Inc()
+		queue := p.requeuePodViaQueueingHint(logger, pInfo, queueAfterBackoff, framework.ScheduleAttemptFailure)
```
`addUnschedulableWithoutQueueingHint` is called when QueueingHints are disabled, so I maybe `requeuePodViaQueueingHint` should be renamed to something like `requeuePodWithStrategy` to not suggest it's a hint?

dom4ha commented on pkg/scheduler/backend/queue/scheduling_queue.go:
Can you fix the function name in the comment of  addUnschedulableWithoutQueueingHint?

dom4ha commented on pkg/scheduler/backend/queue/scheduling_queue.go:
```diff
@@ -1306,14 +1314,14 @@ func (p *PriorityQueue) getBackoffTime(podInfo *framework.QueuedPodInfo) time.Ti
 func (p *PriorityQueue) calculateBackoffDuration(podInfo *framework.QueuedPodInfo) time.Duration {
-	if podInfo.Attempts == 0 {
+	if podInfo.UnschedulableCount == 0 {
 		// When the Pod hasn't experienced any scheduling attempts,
```
s/scheduling attempts/unschedulable verdicts/ ?

sanposhiho commented:
/retitle feat: introduce pInfo.UnschedulableCount to make the backoff calculation more appropriate 
/unhold
implemented the @macsko 's idea ([ref](https://github.com/kubernetes/kubernetes/issues/128744#issuecomment-2488230310)). PTAL again!

sanposhiho commented:
/hold
Ah, I still need a hold to block it from being merged until getting two reviews.

haosdent commented on pkg/scheduler/backend/queue/scheduling_queue.go:
```diff
@@ -771,6 +774,16 @@ func (p *PriorityQueue) AddUnschedulableIfNotPresent(logger klog.Logger, pInfo *
+	} else {
+		// This Pod is rejected by some plugins, not coming back due to unexpected errors (e.g., a network issue)
+		pInfo.UnschedulableCount++
+		// We should reset the error count because the error is not gone.
```
This comment is confusing to me. Because if "the error is not gone" mean we should keep `pInfo.ErrorCount`?
But according to the following changes, should we say some plugins in this scheduling cycle have rejected this pod, we ignore previous unexpected error counts because we prioritize the current unscheduable error.

sanposhiho replied:
Sorry, it's just a mistake; I should've say "the error is gone"

haosdent replied:
Got it, thanks

sanposhiho commented on pkg/scheduler/backend/queue/scheduling_queue.go:
```diff
@@ -771,6 +774,16 @@ func (p *PriorityQueue) AddUnschedulableIfNotPresent(logger klog.Logger, pInfo *
 		return fmt.Errorf("Pod %v is already present in the backoff queue", klog.KObj(pod))
 	}
 
+	if pInfo.UnschedulablePlugins.Union(pInfo.PendingPlugins).Len() == 0 {
```
Yes, we can do that.

dom4ha commented on pkg/scheduler/backend/queue/scheduling_queue.go:
```diff
@@ -771,6 +774,16 @@ func (p *PriorityQueue) AddUnschedulableIfNotPresent(logger klog.Logger, pInfo *
 
+	if len(pInfo.UnschedulablePlugins) == 0 && len(pInfo.PendingPlugins) == 0 {
+		// This Pod came back because of some unexpected errors (e.g., a network issue).
+		pInfo.ErrorCount++
```
Since it's not cumulative count, maybe it should be called `ConsecutiveErrors` or `ConsecutiveErrorsCount`?

dom4ha commented on pkg/scheduler/backend/queue/scheduling_queue.go:
```diff
@@ -771,6 +774,16 @@ func (p *PriorityQueue) AddUnschedulableIfNotPresent(logger klog.Logger, pInfo *
+		pInfo.ErrorCount++
+	} else {
+		// This Pod is rejected by some plugins, not coming back due to unexpected errors (e.g., a network issue)
+		pInfo.UnschedulableCount++
```
True, that's why renaming would make it more straightforward. Each time we remove it from the queue, we increment the backoff (unless the attempt was in effect of a consecutive error)

sanposhiho replied:
I don't get the point overall. 
The backoff time for pods depends on whether the last status was error or unschedulable and also the last last status. (because we have to reset the backoff time if pod was getting error and now got unschedulable, or vise versa) Should we record those at PodInfo so that we can calculate the backoff time at `Pop()`?
Also, in the first place, it still looks conceptually weird to me even after renaming, to calculate a new backoff time at the beginning of a new scheduling cycle even though the pod hasn't failed yet, may or may not fail at the scheduling cycle. How does it make more sense to deliberately calculate the backoff time at Pop(), rather than calculating the backoff time when pods are actually failed and the queue has to the backoff time right before moving the pod into backoffQ?

sanposhiho commented:
@macsko @dom4ha PTAL

macsko commented on pkg/scheduler/backend/queue/backoff_queue_test.go:
```diff
@@ -42,28 +42,35 @@ func TestBackoffQueue_calculateBackoffDuration(t *testing.T) {
+			want:                   32 * time.Nanosecond,
+		},
+		{
+			name:                   "backoff is calculated with ConsecutiveErrorsCount",
```
Add one more test case, where UnschedulableCount > 0 and ConsecutiveErrorsCount > 0

sanposhiho replied:
I changed `UnschedulableCount` to be >0.

macsko commented on pkg/scheduler/backend/queue/scheduling_queue.go:
```diff
@@ -718,7 +718,7 @@ func (p *PriorityQueue) addUnschedulableWithoutQueueingHint(logger klog.Logger,
 	}
 	if p.moveRequestCycle >= podSchedulingCycle || len(rejectorPlugins) == 0 {
-		// Two cases to move a Pod to the active/backoff queue:
+		// Two cases to move a Pod to ctive/backoff queue:
```
Revert the change here:
```suggestion
		// Two cases to move a Pod to the active/backoff queue:
```

macsko commented:
/lgtm

dom4ha commented:
Can you rebase? There are a lot of changes that went through in the meantime.

dom4ha commented on pkg/scheduler/backend/queue/backoff_queue.go:
```diff
@@ -222,14 +222,21 @@ func (bq *backoffQueue) getBackoffTime(podInfo *framework.QueuedPodInfo) time.Ti
+		count = podInfo.ConsecutiveErrorsCount
+	}
+
+	for i := 1; i < count; i++ {
```
The `BackoffExpiration` time is calculated lazy now, so we clear it whenever we change elements that take part in the calculations. So far it were `Attempt` and `Timestamp`, so now we should clear it when we change `UnschedulableCount` and `ConsecutiveErrorsCount`, but not clear it on `Attempt` changes.

dom4ha commented on pkg/scheduler/backend/queue/backoff_queue.go:
```diff
@@ -222,14 +222,21 @@ func (bq *backoffQueue) getBackoffTime(podInfo *framework.QueuedPodInfo) time.Ti
 // based on the number of attempts the pod has made.
 func (bq *backoffQueue) calculateBackoffDuration(podInfo *framework.QueuedPodInfo) time.Duration {
-	if podInfo.Attempts == 0 {
+	if podInfo.UnschedulableCount == 0 && podInfo.ConsecutiveErrorsCount == 0 {
```
Maybe move this check after we calculate `count`, which is the actual replacement of `Attempts`.

dom4ha commented on pkg/scheduler/backend/queue/backoff_queue.go:
```diff
@@ -222,14 +222,21 @@ func (bq *backoffQueue) getBackoffTime(podInfo *framework.QueuedPodInfo) time.Ti
 
 	duration := bq.podInitialBackoff
-	for i := 1; i < podInfo.Attempts; i++ {
+	count := podInfo.UnschedulableCount
```
Please also change `getBackoffTime` and replace `Attempts` there and in the comment. It won't be a simple replacemetn though, so maybe calculate `count` there and pass it to `calculateBackoffDuration` to not count it twice...

dom4ha reviewed (commented):
Unfortunately after rebase, some things has changed because of the lazy calculation, so I have a few comments.

Anyway, our changes to pop from backoff, may greatly increase the number of time we incement attempts (now UnschedulableCount). We may reach max backoff very quickly in some situations. Usually it shouldn't be a problem because of Poping them ASAP, so we should not worry that much, except for errors.

Your change would actually help here if we could calculate errors separately, but not sure if we magage to submit.

sanposhiho commented:
Sorry for a long delay here, updated the PRs @macsko @dom4ha

dom4ha commented on pkg/scheduler/backend/queue/scheduling_queue.go:
```diff
@@ -804,6 +804,18 @@ func (p *PriorityQueue) AddUnschedulableIfNotPresent(logger klog.Logger, pInfo *
+		pInfo.ConsecutiveErrorsCount = 0
+	}
+	// We changed ConsecutiveErrorsCount or UnschedulableCount, and now the calculated backoff time would be different, removing the cached backoff time.
+	pInfo.BackoffExpiration = time.Time{}
```
Can we move setting reseting `pInfo.Timestamp` (from line 825) before we clear `BackoffExpriation`? We'd have all inputs for calculation backoff in one place.

Note that we reset timestamp also in `addUnschedulableWithoutQueueingHint`, which I think's would be not needed, unless I'm missing something.

dom4ha commented on pkg/scheduler/backend/queue/scheduling_queue.go:
```diff
@@ -804,6 +804,18 @@ func (p *PriorityQueue) AddUnschedulableIfNotPresent(logger klog.Logger, pInfo *
+		// This Pod came back because of some unexpected errors (e.g., a network issue).
+		pInfo.ConsecutiveErrorsCount++
+	} else {
+		// This Pod is rejected by some plugins, not coming back due to unexpected errors (e.g., a network issue)
+		pInfo.UnschedulableCount++
+		// We should reset the error count because the error is gone.
```
One more thought, since we do pop from the backoff queue as best effort (scheduler has spare cycle) it may be very frequent (before `BackoffExpiration`).

WDYT about not incrementing `UnschedulableCount` (and not reseting `BackoffExpiration`) when the previous `BackoffExpiration` haven't expired yet? Since pop is best effort, we don't have to penalize such pod, but we should avoid opposite case, when frequent best effort pops bumped `UnschedulableCount` counter too much.

dom4ha replied:
I think you're right, that might be an undesired side effect. Shall we worry about it at all?

The scenario that I was thinking of is that there are many events triggering a pods (qhints) and scheduler is idle. When it gets busy, an affected pod may wait 5min (max backoff). So on one hand we can say that the scheduler is busy anyway and it will pop from backoff once it becomes idle again, but on the other hand, higher priority pod might be the one waiting in unschedulable, which may break what have been just scheduled.

It doesn't sound like a very common case, wdyt?

sanposhiho replied:
I don't agree.
Besides @macsko's point, I believe we shouldn't do that, either way. We should keep the core philosophy in mind; our backoff is the penalty that the pod has wasted one scheduling cycle. We should not commute a sentence even if the scheduler retries the pod frequently. We only do that is when no pod is in activeQ = when retrying those offenders is better than doing nothing.

Is it a bad side effect of https://github.com/kubernetes/enhancements/issues/5142? No, at all. Once the pod gets `Queue` from QHint, _**at that moment**_, it should have a high probability of getting scheduled in the next scheduling cycle. If the pod couldn't make it after all, regardless of whether they went through shorter or longer backoff time, they should be punished.

> we should avoid opposite case, when frequent best effort pops bumped UnschedulableCount counter too much.

It means many cluster events trigger the pod to be moved to backoffQ frequently, but the pod keeps being rejected at the scheduling cycle. 
Putting a longer backoff time for the pod makes complete sense because this means the pod could be _easily_ moved to backoffQ by involved QHints, but got rejected by the scheduling cycle at _all chances._ The pod has got a lot of chances, but it couldn't be scheduled by any of them. That means the reliability of the QHint decision for the pod is quite low. We should give more chances to other pods, we should strictly avoid keep giving frequent chances to this pod because we already know the probability is not high.

macsko replied:
I agree with @sanposhiho points

dom4ha replied:
QHints won't be ever precise due to time constraints, so I still think that high number of update events may bump up the metric and we cannot blame QHints for it but rather a specific cluster behavior.

Anyway, we have pop from backoff feature, so it's not expected pods would wait for the max backoff when we increase it, but rescheduled as soon as possible, so it's hard to predict whether it would be any issue in practice.

sanposhiho commented on pkg/scheduler/backend/queue/backoff_queue_test.go:
```diff
@@ -31,47 +31,94 @@ import (
+			want:                   time.Date(2023, 10, 1, 0, 0, 0, 1, time.UTC),
+		},
+		{
+			name:                   "backoff by UnschedulableCount",
```
done at https://github.com/kubernetes/kubernetes/pull/131714

sanposhiho commented:
Rebased to resolve the conflict and squashed once. The only change I applied is https://github.com/kubernetes/kubernetes/pull/128748#discussion_r2086618335 from @dom4ha

macsko commented on pkg/scheduler/backend/queue/backoff_queue_test.go:
```diff
@@ -31,54 +31,101 @@ import (
+		t.Run(tt.name, func(t *testing.T) {
+			bq := newBackoffQueue(clock.RealClock{}, tt.initialBackoffDuration, tt.maxBackoffDuration, newDefaultQueueSort(), true)
+			if got := bq.getBackoffTime(tt.podInfo); got != tt.want {
+				t.Errorf("backoffQueue.calculateBackoffDuration() = %v, want %v", got, tt.want)
```
```suggestion
				t.Errorf("backoffQueue.getBackoffTime() = %v, want %v", got, tt.want)
```

macsko commented on pkg/scheduler/backend/queue/backoff_queue_test.go:
```diff
@@ -93,30 +140,30 @@ func TestBackoffQueue_popAllBackoffCompleted(t *testing.T) {
-			Timestamp: fakeClock.Now().Add(-2 * time.Second),
-			Attempts:  1,
+			Timestamp:          fakeClock.Now().Add(-2 * time.Second),
+			UnschedulableCount: 1,
```
Probably worth change to `ConsecutiveErrorsCount` for pods that will be in `errorBackoffQ`
```suggestion
			ConsecutiveErrorsCount: 1,
```

macsko commented on pkg/scheduler/backend/queue/backoff_queue_test.go:
```diff
@@ -93,30 +140,30 @@ func TestBackoffQueue_popAllBackoffCompleted(t *testing.T) {
-			Timestamp: fakeClock.Now().Add(time.Second),
-			Attempts:  1,
+			Timestamp:          fakeClock.Now().Add(time.Second),
+			UnschedulableCount: 1,
```
```suggestion
			ConsecutiveErrorsCount: 1,
```

dom4ha commented:
/approve
It looks good from my side, let's wait for lgtm from @macsko

sanposhiho commented:
@macsko Fixed.

sanposhiho commented:
/retest

macsko commented:
/lgtm
/approve

sanposhiho commented:
Got two approvals,
/unhold

sanposhiho commented:
/test pull-kubernetes-e2e-kind
